### ISSUE TITLE
Serverside input transformation

### ISSUE THREAD

**Issue by lulzmachine** [deepstreamIO/deepstream.io#872]
Hello!
We're using ` "deepstream.io": "3.1.0",`.
We've been looking at doing some serverside input transformation, and I'd like to know what the preferred way is. Previously we could set a dataTransform thing on the deepstream server.

The usecase is that we have endusers that make rpc calls to our backend. Right now we have the enduser clients add a "customerId" on each rpc call to make sure we know on the server who made the call. To make sure the customerId really belongs to the user, we check it with a permission handler, like: 

```
module.exports = class CustomPermissionHandler extends EventEmitter {
  canPerformAction (username, message, callback, authData) {
...
        const originalData = JSON.parse(message.data[2].substr(1))
        const isValid = authData.customerId === originalData.customerId
        return callback(null, isValid)
```

But since we do this, and we have the information on server-side, it feels a bit weird that the client has to supply it.

To compare to a REST server, there you would sent a cookie or token, and have the server figure out who the user is. It's not that the user sends who they are and the server verifies it another time. 

So basically we want to be able to add "customerId" on all (or most at least) RPCs coming in from end users. I figured out a way to do it:

```
// here 'echo' is the name of our RPC
    if (message.data[0] === 'echo' && authData.role !== 'backend') {
        const data = JSON.parse(message.data[2].substr(1))
        data.customerId = authData.customerId

        message.data[2] = 'O' + JSON.stringify(data)
        return callback(null, true)
      }
    }
```

This actually works, but feels a bit ugly and relying on onducmented behaviour. So I just thought I'd ask: is this the proper way to do it? If not, then how should it be done?

**Comment by yasserf** [comment]
Hey @lulzmachine 

Amending data currently can only be done in the permission handler. V4 makes it a bit easier by not having all the objects types..

```
// here 'echo' is the name of our RPC
    if (message.name === 'echo' && authData.role !== 'backend') {
        const data = socketWrapper.parseData(message) // part of new api
        delete message.data // forces reserialization when sending
        message.parsedData.customerId = authData.customerId
        return callback(null, true)
      }
    }
```

The goal is to try and minimize the amount of code users can inject into deepstream. We could also reintroduce the RPC transform in a future version as it does seem to be the best way to amend incoming rpc data (not so much with outgoing though).

**Comment by lulzmachine** [comment]
Ok, thanks for the response!
Instead of having transform handlers and permissions handlers, perhaps you could do like react and many HTTP frameworks, in that they let you register middlewares that can do anything.

Something like
```
const customerIdAdder = (req, next) => {
  if (req.authData.role !== 'backend') {
    req.data.customerId = req.authData.customerId
  }
  next()
}

deepstream.use('rpc', 'echo', customerIdAdder)
```
Just a thought

**Comment by lulzmachine** [comment]
@yasserf in your example, what is "socketWrapper" from? Is it available in ds3?

**Comment by yasserf** [comment]
It will be available in ds4. Having typescript allows us to create better/more concrete APIs.

**Comment by yasserf** [comment]
Added this feature to V4 branch, RPC now can send requestorName and requestorClientData alongside the RPC. This just needs to be added to client.

**Comment by yasserf** [comment]
Closing due to inactivity. This feature is part of the spec but not in the RPC callback client itself.
